Added support for AES-ECB to the PSA Crypto implementation

stevew817 opened the pull request:
## Description
This PR implements support for `PSA_ALG_ECB_NO_PADDING` as defined in the PSA Crypto API spec v1.0.0. Feature was requested in #3277.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [X] Tests
- [X] Documentation
- [x] Changelog updated


## Steps to test or reproduce
Tested through the PSA Crypto test suite.

stevew817 commented:
Does this need a changelog entry?

gilles-peskine-arm commented:
I think yes, this counts as a new feature so it should have a changelog entry.

hanno-arm commented on include/mbedtls/psa_util.h:
```diff
@@ -85,6 +85,8 @@ static inline psa_algorithm_t mbedtls_psa_translate_cipher_mode(
     switch( mode )
     {
+        case MBEDTLS_MODE_ECB:
+            return ( PSA_ALG_ECB_NO_PADDING );
```
```suggestion
            return( PSA_ALG_ECB_NO_PADDING );
```

hanno-arm commented on include/psa/crypto_values.h:
```diff
@@ -980,6 +980,15 @@
+ * The underlying block cipher is determined by the key type.
+ *
+ * This symmetric cipher mode can only be used with messages whose lengths
+ * are whole number of blocks for the chosen block cipher.
```
Minor: There's an article missing here.

hanno-arm commented on include/psa/crypto_values.h:
```diff
@@ -980,6 +980,15 @@
+ *
+ * The underlying block cipher is determined by the key type.
+ *
+ * This symmetric cipher mode can only be used with messages whose lengths
+ * are whole number of blocks for the chosen block cipher.
```
```suggestion
 * This symmetric cipher mode can only be used with messages whose lengths
 * are a multiple of the block size of the chosen block cipher.
```

stevew817 replied:
Copied from the PSA specification. I'd rather have the specification updated before replicating any changes here.

gilles-peskine-arm replied:
I've submitted a patch to change the sentence to

> This symmetric cipher mode can only be used with messages whose lengths are a multiple of the block size of the chosen block cipher.

hanno-arm commented on library/psa_crypto.c:
```diff
@@ -3746,7 +3749,11 @@ static psa_status_t psa_cipher_init( psa_cipher_operation_t *operation,
     operation->key_set = 0;
     operation->iv_set = 0;
-    operation->iv_required = 1;
+    if( alg == PSA_ALG_ECB_NO_PADDING ) {
```
```suggestion
    if( alg == PSA_ALG_ECB_NO_PADDING ) 
    {
```

hanno-arm commented on library/psa_crypto.c:
```diff
@@ -3746,7 +3749,11 @@ static psa_status_t psa_cipher_init( psa_cipher_operation_t *operation,
-    operation->iv_required = 1;
+    if( alg == PSA_ALG_ECB_NO_PADDING ) {
+        operation->iv_required = 0;
+    } else {
```
```suggestion
    } else 
    {
```

hanno-arm commented on library/psa_crypto.c:
```diff
@@ -3837,7 +3844,7 @@ static psa_status_t psa_cipher_setup( psa_cipher_operation_t *operation,
     operation->block_size = ( PSA_ALG_IS_STREAM_CIPHER( alg ) ? 1 :
                               PSA_BLOCK_CIPHER_BLOCK_SIZE( slot->attr.type ) );
-    if( alg & PSA_ALG_CIPHER_FROM_BLOCK_FLAG )
+    if( alg & PSA_ALG_CIPHER_FROM_BLOCK_FLAG && alg != PSA_ALG_ECB_NO_PADDING )
```
Suggestion: Add brackets for clarity.
```suggestion
    if( ( alg & PSA_ALG_CIPHER_FROM_BLOCK_FLAG ) != 0 && 
        alg != PSA_ALG_ECB_NO_PADDING )
```

hanno-arm commented on library/psa_crypto.c:
```diff
@@ -3991,12 +3998,14 @@ psa_status_t psa_cipher_finish( psa_cipher_operation_t *operation,
-        operation->ctx.cipher.unprocessed_len != 0 )
+    if( operation->ctx.cipher.unprocessed_len != 0 )
     {
+        if( operation->alg == PSA_ALG_ECB_NO_PADDING ||
+            ( operation->alg == PSA_ALG_CBC_NO_PADDING &&
+              operation->ctx.cipher.operation == MBEDTLS_ENCRYPT ) ) {
```
```suggestion
        if( operation->alg == PSA_ALG_ECB_NO_PADDING ||
            ( operation->alg == PSA_ALG_CBC_NO_PADDING &&
              operation->ctx.cipher.operation == MBEDTLS_ENCRYPT ) ) 
        {
```

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3347,7 +3347,11 @@ void cipher_encrypt( int alg_arg, int key_type_arg,
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
+    if( iv->len > 0 )
+    {
+      PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```
```suggestion
    {
         PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3410,7 +3414,11 @@ void cipher_encrypt_multipart( int alg_arg, int key_type_arg,
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
+    if( iv->len > 0 )
+    {
+      PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```
```suggestion
    {
         PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3479,7 +3487,9 @@ void cipher_decrypt_multipart( int alg_arg, int key_type_arg,
                                           handle, alg ) );
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
+    if( iv->len > 0 ) {
+      PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```
```suggestion
    if( iv->len > 0 ) 
    {
        PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```

hanno-arm replied:
Minor: The brace `{` is still misplaced here and at other occasions.

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3546,7 +3556,9 @@ void cipher_decrypt( int alg_arg, int key_type_arg,
                                           handle, alg ) );
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
+    if( iv->len > 0 ) {
+      PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```
```suggestion
    if( iv->len > 0 ) 
    {
        PSA_ASSERT( psa_cipher_set_iv( &operation, iv->x, iv->len ) );
```

hanno-arm replied:
Same here.

stevew817 replied:
Done now, sorry for the oversight.

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3613,9 +3625,11 @@ void cipher_verify_output( int alg_arg, int key_type_arg,
-    PSA_ASSERT( psa_cipher_generate_iv( &operation1,
-                                        iv, iv_size,
-                                        &iv_length ) );
+    if( alg != PSA_ALG_ECB_NO_PADDING ) {
+      PSA_ASSERT( psa_cipher_generate_iv( &operation1,
```
```suggestion
    if( alg != PSA_ALG_ECB_NO_PADDING ) 
    {
        PSA_ASSERT( psa_cipher_generate_iv( &operation1,
```

hanno-arm replied:
Same here.

stevew817 replied:
Done now, sorry for the oversight.

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3635,8 +3649,11 @@ void cipher_verify_output( int alg_arg, int key_type_arg,
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation2,
-                                   iv, iv_length ) );
+    if( iv_length > 0 ) {
+      PSA_ASSERT( psa_cipher_set_iv( &operation2,
```
```suggestion
    if( iv_length > 0 ) 
    {
        PSA_ASSERT( psa_cipher_set_iv( &operation2,
```

hanno-arm replied:
Same here.

stevew817 replied:
Done now, sorry for the oversight.

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3698,9 +3715,12 @@ void cipher_verify_output_multipart( int alg_arg,
-    PSA_ASSERT( psa_cipher_generate_iv( &operation1,
-                                        iv, iv_size,
-                                        &iv_length ) );
+    if( alg != PSA_ALG_ECB_NO_PADDING ) {
+      PSA_ASSERT( psa_cipher_generate_iv( &operation1,
```
```suggestion
    if( alg != PSA_ALG_ECB_NO_PADDING ) 
    {
        PSA_ASSERT( psa_cipher_generate_iv( &operation1,
```

hanno-arm replied:
Same here.

stevew817 replied:
Done now, sorry for the oversight.

hanno-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -3730,8 +3750,10 @@ void cipher_verify_output_multipart( int alg_arg,
 
-    PSA_ASSERT( psa_cipher_set_iv( &operation2,
-                                   iv, iv_length ) );
+    if( iv_length > 0 ) {
+      PSA_ASSERT( psa_cipher_set_iv( &operation2,
```
```suggestion
    if( iv_length > 0 ) 
    {
        PSA_ASSERT( psa_cipher_set_iv( &operation2,
```

hanno-arm replied:
Same here.

stevew817 replied:
Done now, sorry for the oversight.

stevew817 commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -1122,6 +1126,10 @@ PSA cipher: bad order function calls
+PSA symmetric encrypt: AES-ECB, 16 bytes, good
+depends_on:MBEDTLS_AES_C
+cipher_encrypt:PSA_ALG_ECB_NO_PADDING:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":"":"6bc1bee22e409f96e93d7e117393172a":"3ad77bb40d7a3660a89ecaf32466ef97":PSA_SUCCESS
+
```
3DES isn't in scope for this PR?

gilles-peskine-arm replied:
This PR makes Mbed TLS recognize `PSA_ALG_ECB_NO_PADDING`, so it should test the algorithm with at least a selection of block ciphers. 3DES and other block ciphers should either work correctly or return `NOT_SUPPORTED`. I requested a test for 3DES and not for e.g. CAMELLIA because CAMELLIA is very likely to work if AES does, whereas DES and 3DES could be incorrect due to a misuse of a hard-coded 16 instead of the block size somewhere.

gilles-peskine-arm commented on ChangeLog.d/add-aes-ecb-to-psa.txt:
```diff
@@ -0,0 +1,2 @@
+Features
+   * Added support for AES-ECB to the PSA Crypto cipher API.
```
Actually, it should be support for ECB. Not specifically AES-ECB, which isn't a specific thing in the API.

stevew817 commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -1210,6 +1226,10 @@ PSA symmetric encrypt/decrypt: AES-CTR
+PSA symmetric encryption multipart: AES-ECB, 16+16 bytes
+depends_on:MBEDTLS_AES_C
+cipher_encrypt_multipart:PSA_ALG_ECB_NO_PADDING:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":"":"6bc1bee22e409f96e93d7e117393172a5434f378a597bcef1389318c7fc865ef":16:16:16:"3ad77bb40d7a3660a89ecaf32466ef9755ed5e9e066820fa52c729886d18854c"
+
```
There weren't any negative test cases being done with _multipart tests. Is it really up to me to introduce that here?

gilles-peskine-arm replied:
A prior test gap is not a justification for expanding the test gap in general. But ok, let's [defer negative testing for multipart operations](https://github.com/ARMmbed/mbedtls/issues/3607) since it's a preexisting lack of test code. Do please add test with input that isn't split on a block boundary, which was already done for CBC.

stevew817 replied:
Sure (push incoming for the non-block-multiple operations).

stevew817 commented:
@hanno-arm I made your suggested style fixed and changed the wording to what @gilles-peskine-arm noted the spec would be updated to.

Please review the last commit as well. I added the tests as requested by @gilles-peskine-arm and had to implement manual reassembling of multipart input to block-size chunks, since the PSA API does not match the mbed TLS cipher API for ECB.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -3975,9 +3975,83 @@ psa_status_t psa_cipher_update( psa_cipher_operation_t *operation,
+                *output_length += internal_output_length;
+            }
+
+            size_t blocks = input_length / operation->block_size;
+            for( ; blocks > 0; blocks-- )
```
Please rewrite the loop to use something like
```
for( size_t n = input_length; n >= operation->block_size; n -= operation->block_size )
```
avoiding division by the block size.

We avoid division with a non-constant in symmetric cryptography code because on some platforms without hardware division, it requires pulling in a library function. This is admittedly not well documented. There is a configuration option `MBEDTLS_NO_UDBL_DIVISION` which informs the choice of word size in the bignum module, and no part of the crypto library should use division on a larger type (apart from division by a constant). A platform could have a 64-bit `size_t` and a 32-bit `mbedtls_mpi_uint` and no 64-bit hardware division, although I'd expect that on most such platforms linking a software division function wouldn't be a problem. We have no such platform on our CI (we test the effectiveness of `MBEDTLS_NO_UDBL_DIVISION` on a 32-bit arm build, with a 32-bit `size_t`). I don't think it's worth taking the risk, even if it's only in a somewhat “exotic” platform. Here the code can easily be rewritten.

stevew817 replied:
Done.

gilles-peskine-arm replied:
Looking further, I see that there was already a division by the block size in `psa_crypto.c`, and [a few other similar cases in the code base](https://github.com/ARMmbed/mbedtls/issues/3654). I don't think this was intended, but some have been around for years, so I don't think they bother anyone. I know that 64-bit division on 32-bit platforms is an actual problem that some users of Mbed TLS have faced, but division by `size_t` apparently isn't.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -3975,9 +3975,83 @@ psa_status_t psa_cipher_update( psa_cipher_operation_t *operation,
-                                 input_length, output, output_length );
-    status = mbedtls_to_psa_error( ret );
+    if( operation->alg == PSA_ALG_ECB_NO_PADDING )
+    {
```
This is a very big block of code. Please split off either this block, or the `if( expected_output_size > 0 )` block, as a separate function.

stevew817 replied:
Done.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -1326,19 +1382,19 @@ PSA symmetric encryption multipart: AES-CTR, 11+5 bytes [#2]
 cipher_decrypt_multipart:PSA_ALG_CTR:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":"2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a":"6bc1bee22e409f96e93d7e117393172a":11:11:5:"8f9408fe80a81d3e813da3c7b0b2bd32"
 
-PSA symmetric encryption multipart: AES-CTR, 16+16 bytes [#2]
+PSA symmetric decryption multipart: AES-CTR, 16+16 bytes [#2]
```
Thanks for noticing and fixing the copypasta. Now that it's fixed, the `[#2]` suffix, and `[#1]` on the actual encryption test cases, can be removed. The suffixes `[#NUMBER]` were added automatically to make test case descriptions unique for reporting purpose, and fixing the duplicates (sometimes due to copypasta as here, sometimes due to descriptions that aren't precise enough) is an [ongoing task](https://github.com/ARMmbed/mbed-crypto/issues/271).

stevew817 replied:
Sure.

stevew817 replied:
Done.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -1142,6 +1150,10 @@ PSA symmetric encrypt: AES-CBC-PKCS#7, 15 bytes, good
 
+PSA symmetric encrypt: AES-ECB, input too short (15 bytes)
+depends_on:MBEDTLS_AES_C
+cipher_encrypt:PSA_ALG_ECB_NO_PADDING:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":"":"6bc1bee22e409f96e93d7e11739317":"3ad77bb40d7a3660a89ecaf32466ef":PSA_ERROR_INVALID_ARGUMENT
```
```suggestion
cipher_encrypt:PSA_ALG_ECB_NO_PADDING:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf7158809cf4f3c":"":"6bc1bee22e409f96e93d7e11739317":"":PSA_ERROR_INVALID_ARGUMENT
```
The test passes either way since it ignores the output when an error occurs, but I find including non-empty output that doesn't actually matter misleading.

stevew817 replied:
Sure.

stevew817 replied:
Done.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -3927,15 +3938,102 @@ psa_status_t psa_cipher_set_iv( psa_cipher_operation_t *operation,
+
+    status = PSA_SUCCESS;
+
+exit:
```
Since there's no cleanup, it would be more readable to use `return( status )` instead of `goto exit` (and `return( whatever );` instead of `status = whatever; goto exit;`). I don't consider it a blocker.

stevew817 commented:
@hanno-arm I'd appreciate it if you could give your thumbs up now that all your requested changes have been incorporated. This PR has been open for over two months already.

hanno-arm commented on library/psa_crypto.c:
```diff
@@ -3964,9 +4062,24 @@ psa_status_t psa_cipher_update( psa_cipher_operation_t *operation,
-    status = mbedtls_to_psa_error( ret );
+    if( operation->alg == PSA_ALG_ECB_NO_PADDING )
+    {
+        /* mbedtls_cipher_update has an API inconsistency: it will only
```
After moving the bulk of the code to `psa_cipher_update_ecb_internal()`, this comment seems misplaced. Not a blocker.

hanno-arm reviewed (approved):
Apologies for the delay, @stevew817, and thank you for making the requested changes. LGTM

bensze01 commented:
All the mbedTLS tests have passed (at the time of writing, only the MbedOS tests are running).
